### ISSUE TITLE
mAP problems in train.py and test.py

### ISSUE THREAD

**Issue by yinyuhangi** [ultralytics/yolov5#1397]
## ❔Question
The mAP of the last epoch during training was 0.465.
![image](https://user-images.githubusercontent.com/46810994/99146567-52cbf880-26b4-11eb-8c70-7c20f868ff25.png)
But when I run test.py, the mAP is only 0.459. Why? I'm using the same test set.
![image](https://user-images.githubusercontent.com/46810994/99146583-8ad33b80-26b4-11eb-82dc-071cab62f4b1.png)
Looking forward to your reply!

**Comment by glenn-jocher** [comment]
@yinyuhangi if you look at the test dataloaders in train.py and test.py, the only difference is that the test.py dataloader passes a pad=0.5 argument, which pads the image with a half stride on each side, producing improved mAP on COCO. You might try matching the two for identical performance.

https://github.com/ultralytics/yolov5/blob/9b0f6e33ee6affb648dd9390739e340e91d6bb70/test.py#L84-L91

**Comment by yinyuhangi** [comment]
Thank you very much. I changed the pad to 0.0 and the mAP of test.py to 0.461, but it still failed to reach 0.465 in train.py. @glenn-jocher

**Comment by glenn-jocher** [comment]
@yinyuhangi well then the only difference remaining would be the batch size. Padding constraints are dynamically updated as a function of batch size, so you should figure out what batch size you used for training and duplicate that in test.py.

**Comment by isaacpascal** [comment]
Hello @glenn-jocher,
Actually, this repo is great.

I also have same probem. My training batch-size is 16 in train.py, and 32 in test.py,  actually they are default. Now, I have changed pad 0.0 in test.py, still map and other metrics are different. 

Changing pad 0.5 to 0.0 reduces the map. 

Also, I realized that the best.pt model sometimes gives a lower map than the next last.pt (even last.pt map is lower than best.pt). As if the best weights are not stored in best.pt. Maybe, everything is related to this question...
Thanks.

**Comment by glenn-jocher** [comment]
@isaacpascal best.pt is saved when a new maximum fitness epoch is observed. Fitness is defined as a weighted combination of metrics in utils/metrics.py.

If you believe there is a bug please raise a bug report with links to a **reproducible** example in a colab notebook. Thanks!

**Comment by gizemtanriver** [comment]
I am having a similar issue. During training, the results of the best epoch are:
```
 Epoch   gpu_mem       box       obj       cls     total   targets  img_size
     23/79     5.86G   0.04022   0.02975         0   0.06997         6       512: 100% 67/67 [00:30<00:00,  2.17it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100% 8/8 [00:00<00:00,  9.16it/s]
                 all          59          60       0.454       0.867       0.734       0.355
```
But when I try to use the best.pt with detect.py, it gives worse results in all metrics except precision. The fitness is defined as the weighted average of the metrics, weights being [0.0, 0.0, 0.1, 0.9] respectively. So I don't understand why best.pt is not keeping my best weights from training.
```
Namespace(augment=False, batch_size=32, conf_thres=0.001, data='datasets/lesion/lesion_data.yaml', device='', exist_ok=False, img_size=512, iou_thres=0.65, name='val_results', project='runs/train/lesion_yolov5x_defaultanchor_512_bs8_evolved100gen_CVsplit0', save_conf=False, save_json=False, save_txt=False, single_cls=True, task='val', verbose=True, weights=['runs/train/lesion_yolov5x_defaultanchor_512_bs8_evolved100gen_CVsplit0/bin/bash/weights/best.pt'])
Fusing layers... 
                 all          59          60       0.505       0.767        0.63       0.342
Speed: 102.1/1.4/103.5 ms inference/NMS/total per 512x512 image at batch-size 32
Results saved to runs/train/lesion_yolov5x_defaultanchor_512_bs8_evolved100gen_CVsplit0/val_results6
Using torch 1.7.0+cu101 CUDA:0 (Tesla K80, 11441MB)
```
I had done the same experiment with the previous v2 of the repo with all the same settings, and I don't experience this issue with v2. I am getting very similar results between the best epoch during training and at test time with detect.py using best.pt.

**Comment by glenn-jocher** [comment]
@gizemtanriver This issue seems to lack the minimum requirements for a proper response, or is insufficiently detailed for us to help you. Please note that most technical problems are due to:

- **Your modified or out-of-date code**. If your issue is not reproducible in a new `git clone` version of this repo we can not debug it. Before going further run this code and verify your issue persists:
```bash
$ git clone https://github.com/ultralytics/yolov5 yolov5_new  # clone latest
$ cd yolov5_new
$ python detect.py  # verify detection

# CODE TO REPRODUCE YOUR ISSUE HERE
```

- **Your custom data.** If your issue is not reproducible in one of our 3 common datasets ([COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), or [VOC](https://github.com/ultralytics/yolov5/blob/master/data/voc.yaml)) we can not debug it. Visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for guidelines on training your custom data. Examine `train_batch0.jpg` and `test_batch0.jpg` for a sanity check of your labels and images.

- **Your environment.** If your issue is not reproducible in one of the verified environments below we can not debug it. If you are running YOLOv5 locally, verify your environment meets all of the [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies specified below. If in doubt, download Python 3.8.0 from https://www.python.org/, create a new [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), and then install requirements.

If none of these apply to you, we suggest you close this issue and raise a new one using the **Bug Report template**, providing screenshots and **minimum viable code to reproduce your issue**. Thank you!


## Requirements

Python 3.8 or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including `torch>=1.6`. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)

## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.
